scripts/xdcc_autoget.pl

```diff
@@ -262,162 +301,168 @@ sub ag_getmsg		#runs when bot sends privmsg. Avoid talking to bots to keep this
+		push(@{$totags[$botcounter]}, Irssi::timeout_add_once($botdelay * 1000, sub { if (!$downloadflag[$botcounter]) { ag_packrequest($botcounter); } } , []));
 	}
 }
 
 sub ag_opendcc	#runs on DCC recieve init
 {
-	&Irssi::signal_continue;
+	Irssi::signal_continue;
 	my ($gdcc) = @_;	#current pack
 	my $botname = $gdcc->{'nick'};
 	my $filename = $gdcc->{'arg'};
+	my $filedownloadflag = 0;
+	
+	$filename =~ tr/[ ']/[__]/;
 	$botname =~ tr/[A-Z]/[a-z]/;
-	$bots[$botcounter] =~ tr/[A-Z]/[a-z]/;
+	
+	foreach my $file (@filenames){ if ($file eq $filename){ $filedownloadflag = 1; }}
 
-	if ($botname eq $bots[$botcounter])	#if it's our bot, let user know, and stop any further AG pack requests until finished
+	my $botcounter = 0;	
+	foreach my $bot (@bots)
 	{
-		Irssi::signal_add("dcc destroyed", "ag_closedcc");
-		Irssi::signal_remove("dcc request", "ag_opendcc");		#stops any suplicate sends (there should only ever be one)
-		Irssi::signal_remove("message irc notice", "ag_getmsg");
-		$getmsgflag = 0;
-
-		&ag_remtimeouts;
-		$dccflag = 0;
-		$downloadflag = 1;
-		$statusbarmessage = "Downloading $filename from $botname, #$packs[$packcounter]"; 
-		foreach my $n (@finished)		#don't redownload finished packs
+		$bot =~ tr/[A-Z]/[a-z]/;
+		if ($botname eq $bot and !$filedownloadflag)	#if it's our bot and the file is not already being downloaded, let user know, and stop any further AG pack requests until finished
 		{
-			if ($n eq $gdcc->{'arg'})	#if file already downloaded, cancel
+			ag_remtimeouts($botcounter);	#stop any other skips
+			$getmsgflag[$botcounter] = 0;
+						
+			$downloadflag[$botcounter] = 1;
+			foreach my $n (@finished)		#don't redownload finished packs
 			{
-				Irssi::signal_remove("dcc destroyed", "ag_closedcc");
-				$gdcc->close;
-				$gdcc->{'skipped'} = $gdcc->{'size'};
-				$gdcc->{'transfd'} = $gdcc->{'size'};
-				ag_closedcc($gdcc);				
+				if ($n eq $gdcc->{'arg'})	#if file already downloaded, cancel
+				{
+					ag_message("msg $bots[$botcounter] $cancelprefix" );
+					ag_addfinished($gdcc->{'arg'}, $botcounter);
+					$skipunfinishedflag[$botcounter] = 1;
+					$gdcc->close;
+				}
+				last if ($n eq $gdcc->{'arg'});
 			}
-			last if ($n eq $gdcc->{'arg'});
+			push(@filenames, $filename);
 		}
+		elsif ($botname eq $bot and $filedownloadflag)	#don't download packs that are being downloaded by other bots
+		{
+			ag_message("msg $bots[$botcounter] $cancelprefix" );
+			$skipunfinishedflag[$botcounter] = 1;
+			$gdcc->close;
+		}
+		$botcounter++;
 	}
 }
 
 sub ag_skip
 {
-#	Irssi::print "AG | SKIP $msgflag $episodicflag $episode $#packs $packcounter $#terms $termcounter $#bots $botcounter"; 
-	&ag_remtimeouts;	#stop any other skips
-	$reqpackflag = 0;		#allow pack requests now that transfer is finished
+	my($botcounter) = @_;
+	my @packlist = ();
+	try {my @packlist = @{$packs[$botcounter]};}	#workaround for @{} being dumb if an array inside of an array is empty 
```

Comment:
another common code which also looks ugly and could be used instead of try/catch: 
`@{$packs[$botcounter]//[]}`
that translates to: if `$packs[$botcounter]` is undef, use `[]` instead (which is an anonymous empty array reference)